File: sdk-go/dephealth/check_result.go
package dephealth

// StatusCategory is a typed alias for status category string values.
// It wraps the status constants used by the error classification system
// and the app_dependency_status metric.
type StatusCategory string

const (
	// StatusOK means the health check succeeded.
	StatusOK StatusCategory = "ok"
	// StatusTimeout means the health check exceeded its deadline.
	StatusTimeout StatusCategory = "timeout"
	// StatusConnectionError means the connection to the dependency failed.
	StatusConnectionError StatusCategory = "connection_error"
	// StatusDNSError means DNS resolution failed for the dependency host.
	StatusDNSError StatusCategory = "dns_error"
	// StatusAuthError means authentication with the dependency failed.
	StatusAuthError StatusCategory = "auth_error"
	// StatusTLSError means a TLS handshake or certificate error occurred.
	StatusTLSError StatusCategory = "tls_error"
	// StatusUnhealthy means the dependency responded but reported unhealthy status.
	StatusUnhealthy StatusCategory = "unhealthy"
	// StatusError means an unclassified error occurred during the health check.
	StatusError StatusCategory = "error"
	// StatusUnknown is used only for HealthDetails API before the first check completes.
	StatusUnknown StatusCategory = "unknown"
)

// AllStatusCategories contains the 8 status categories used for the
// app_dependency_status enum-pattern gauge. StatusUnknown is excluded
// as it is only used for the HealthDetails API.
var AllStatusCategories = []StatusCategory{
	StatusOK,
	StatusTimeout,
	StatusConnectionError,
	StatusDNSError,
	StatusAuthError,
	StatusTLSError,
	StatusUnhealthy,
	StatusError,
}

// CheckResult holds the classification of a health check outcome.
type CheckResult struct {
	Category StatusCategory // One of Status* constants.
	Detail   string         // Specific detail value (e.g. "http_503", "grpc_not_serving").
}

// ClassifiedError is an interface for errors that carry status classification.
// Health checkers can return errors implementing this interface to provide
// precise status category and detail for the app_dependency_status metrics.
type ClassifiedError interface {
	error
	StatusCategory() StatusCategory
	StatusDetail() string
}

// ClassifiedCheckError is a concrete error type that implements ClassifiedError.
// Checkers use this to return classified errors.
type ClassifiedCheckError struct {
	Category StatusCategory
	Detail   string
	Cause    error
}

// Error returns the error message, delegating to Cause if present.
func (e *ClassifiedCheckError) Error() string {
	if e.Cause != nil {
		return e.Cause.Error()
	}
	return e.Detail
}

// Unwrap returns the underlying cause for use with errors.Is/As.
func (e *ClassifiedCheckError) Unwrap() error {
	return e.Cause
}

// StatusCategory returns the status category for this error.
func (e *ClassifiedCheckError) StatusCategory() StatusCategory {
	return e.Category
}

// StatusDetail returns the detail value for this error.
func (e *ClassifiedCheckError) StatusDetail() string {
	return e.Detail
}

File: sdk-python/dephealth/check_result.py
"""Error classification for health check results."""

from __future__ import annotations

import socket
import ssl
from dataclasses import dataclass

# Status category constants.
STATUS_OK = "ok"
STATUS_TIMEOUT = "timeout"
STATUS_CONNECTION_ERROR = "connection_error"
STATUS_DNS_ERROR = "dns_error"
STATUS_AUTH_ERROR = "auth_error"
STATUS_TLS_ERROR = "tls_error"
STATUS_UNHEALTHY = "unhealthy"
STATUS_ERROR = "error"
STATUS_UNKNOWN = "unknown"

ALL_STATUS_CATEGORIES = (
    STATUS_OK,
    STATUS_TIMEOUT,
    STATUS_CONNECTION_ERROR,
    STATUS_DNS_ERROR,
    STATUS_AUTH_ERROR,
    STATUS_TLS_ERROR,
    STATUS_UNHEALTHY,
    STATUS_ERROR,
)


@dataclass(frozen=True)
class CheckResult:
    """Classification of a health check outcome."""

    category: str
    detail: str


def classify_error(err: BaseException | None) -> CheckResult:
    """Classify a check error into a status category and detail.

    Classification chain:
    1. CheckError with status_category/status_detail properties
    2. Standard library error types (TimeoutError, socket errors, SSL errors)
    3. Fallback → error/error
    """
    if err is None:
        return CheckResult(category=STATUS_OK, detail=STATUS_OK)

    # 1. Classified check errors (imports deferred to avoid circular import).
    from dephealth.checker import (
        CheckAuthError,
        CheckConnectionRefusedError,
        CheckDnsError,
        CheckError,
        CheckTimeoutError,
        CheckTlsError,
        UnhealthyError,
    )

    if isinstance(err, CheckError) and err.status_category != STATUS_ERROR:
        return CheckResult(category=err.status_category, detail=err.status_detail)

    # Check typed error subclasses.
    if isinstance(err, CheckTimeoutError):
        return CheckResult(category=STATUS_TIMEOUT, detail="timeout")
    if isinstance(err, CheckConnectionRefusedError):
        return CheckResult(category=STATUS_CONNECTION_ERROR, detail="connection_refused")
    if isinstance(err, CheckDnsError):
        return CheckResult(category=STATUS_DNS_ERROR, detail="dns_error")
    if isinstance(err, CheckAuthError):
        return CheckResult(category=STATUS_AUTH_ERROR, detail="auth_error")
    if isinstance(err, CheckTlsError):
        return CheckResult(category=STATUS_TLS_ERROR, detail="tls_error")
    if isinstance(err, UnhealthyError):
        return CheckResult(category=STATUS_UNHEALTHY, detail=err.status_detail)

    # 2. Platform error detection.
    if isinstance(err, TimeoutError):
        return CheckResult(category=STATUS_TIMEOUT, detail="timeout")
    if isinstance(err, asyncio_timeout_error()):
        return CheckResult(category=STATUS_TIMEOUT, detail="timeout")
    if isinstance(err, socket.gaierror):
        return CheckResult(category=STATUS_DNS_ERROR, detail="dns_error")
    if isinstance(err, ConnectionRefusedError):
        return CheckResult(category=STATUS_CONNECTION_ERROR, detail="connection_refused")
    if isinstance(err, ConnectionError):
        return CheckResult(category=STATUS_CONNECTION_ERROR, detail="connection_refused")
    if isinstance(err, ssl.SSLError):
        return CheckResult(category=STATUS_TLS_ERROR, detail="tls_error")
    if isinstance(err, ssl.SSLCertVerificationError):
        return CheckResult(category=STATUS_TLS_ERROR, detail="tls_error")

    # Check wrapped exceptions.
    cause = getattr(err, "__cause__", None) or getattr(err, "__context__", None)
    if cause is not None and cause is not err:
        inner = classify_error(cause)
        if inner.category != STATUS_ERROR:
            return inner

    # 3. Fallback.
    return CheckResult(category=STATUS_ERROR, detail="error")


def asyncio_timeout_error() -> type:
    """Return asyncio.TimeoutError (same as TimeoutError in 3.11+)."""
    import asyncio

    return asyncio.TimeoutError


__all__ = [
    "ALL_STATUS_CATEGORIES",
    "STATUS_AUTH_ERROR",
    "STATUS_CONNECTION_ERROR",
    "STATUS_DNS_ERROR",
    "STATUS_ERROR",
    "STATUS_OK",
    "STATUS_TIMEOUT",
    "STATUS_TLS_ERROR",
    "STATUS_UNHEALTHY",
    "STATUS_UNKNOWN",
    "CheckResult",
    "classify_error",
]
